Modify do_action so default ctrl-click opens tab

azul opened the pull request:
Rebase and whitespace fix of #16530 

New pull request to see if build succeeds.

azul commented:
tests are failing as they were before.
This error seems to be related to the changes:
```
undefined is not an object (evaluating 'window.FileActions.currentFile = item.find('td')') in apps/files/js/filelist.js (line 933)
```

azul commented on apps/files/js/filelist.js:
```diff
@@ -933,6 +926,28 @@
+			var item = this.$fileList.children().filterAttr('data-id', '' + id);
+			var filename = item.attr('data-file');
+			// also set on global object for legacy apps
+			window.FileActions.currentFile = item.find('td');
```
This is the line triggering the `undefined is not an Object` error.
I suspect that item is not set for some reason.

azul replied:
looks more like `window.FileActions ` does not exist anymore at all.
Dropping this.

azul commented:
Pushed a change that will hopefully help with the tests. I noticed some other strange behavior though...
shift-clicking any file will select all files up to that one plus the last one in the list.
shift-clicking any file in my file browser will select only that file and then start selecting from there with the next shift-click.

I also commented on the initial issue since i don't quite agree with the goal here. But whatever... it's an exercise for now.

jancborchardt reviewed (approved):
@azul I commented in the issue to clarify: Yes, ctrl-click should be the "Open in new tab" action. :)

azul committed:
refactor: fileActions.getCurrentDefaultFileAction()

fileActions.getCurrentDefaultFileAction() returns the default file action
for the currently selected file.

There were a number of places querying for the mime, type and permissions
of that file first to then query for the default action.

Signed-off-by: Azul <azul@riseup.net>

azul committed:
Allow opening files for editing in new tabs

In a file list files with a default action
have an href that will trigger the action.
This way ctrl-click and middle button click open the default action in a new tab.

In order to achieve this a new param `openfile` was introduced to the files app.
It will make the files app trigger the default action for the file in question.
This also allows linking to file content rather than just the details display.

Introduce fileList.getDefaultActionUrl()
to create a link with that param set.
It's overwritten in the trashbin fileList
so that anchors continue to have `#` as a href.

Fix the link generation for subfolders of public shares:
58a87d0 was the last commit that touched the linkTo function in public.js.
It included the params as arguments to the generateUrl function.
Turns out this completely ignores the dir parameter now.
The inclusion was reverted in other places
so revert it here as well.
Also change `dir` to `path` in the param as that is respected
when following the link.

Add Test for the new link url for files with default action.
Remove test for multiple selects with ctrl-click
as that is not what we are doing anymore.

Signed-off-by: Azul <azul@riseup.net>

azul commented on apps/files_sharing/js/public.js:
```diff
@@ -204,10 +204,6 @@ OCA.Sharing.PublicApp = {
 				var $tr = OCA.Files.FileList.prototype._createRow.apply(this, arguments);
 				if (hideDownload === 'true') {
 					this.fileActions.currentFile = $tr.find('td');
-					var mime = this.fileActions.getCurrentMimeType();
-					var type = this.fileActions.getCurrentType();
-					var permissions = this.fileActions.getCurrentPermissions();
-					var action = this.fileActions.getDefault(mime, type, permissions);
```
none of these were used anywhere.

azul commented on apps/files_trashbin/tests/js/filelistSpec.js:
```diff
@@ -180,7 +180,8 @@ describe('OCA.Trashbin.FileList tests', function () {
 			expect($tr.attr('data-mtime')).toEqual('11111000');
-			expect($tr.find('a.name').attr('href')).toEqual('#');
+			expect($tr.find('a.name').attr('href'))
+				.toEqual(OC.getRootPath() + '/index.php/apps/files?view=trashbin&dir=/&editfile=1');
```
I'm not sure if this is the desired behavior.
This is in the trashcan. The href used to be `#`.
It looks like now one could open files in the trashcan in another tab. However this does not happen for other reasons.

- [ ] make the trashcan url a `#` again

azul commented:
Somehow this seems to break the right click menu for me.

azul commented:
Failing tests seem to be the same as in master.
I am also not seeing the right click menu anymore in master. So i also think that is unrelated. Hope that's an actual change... never liked it anyway :wink:
@juliushaertl @gary-kim I think this one is ready to be merged.

gary-kim commented:
> I am also not seeing the right click menu anymore in master.

The right click menu is added by the [files_rightclick](https://github.com/nextcloud/files_rightclick). I you want to test it on current master with right click, you will need to add the app and enable it.

gary-kim reviewed (commented):
The action that occurs from clicking on files to download seems to be broken now. Could you take a look at that?

azul commented:
> The right click menu is added by the [files_rightclick](https://github.com/nextcloud/files_rightclick) app. If you want to test it on current master with right click, you will need to add the app and enable it.

Ahh.. got it. thanks! :smile:

azul commented:
> The action that occurs from clicking on files to download seems to be broken now. Could you take a look at that?

Should be fixed.

This is getting close. I would like to add a test for the new urls to make sure they actually trigger the action. Will probably look into that tomorrow.

gary-kim commented on apps/files_sharing/js/public.js:
```diff
@@ -113,8 +113,8 @@ OCA.Sharing.PublicApp = {
-				if (typeof action !== 'undefined') {
+				var action = FileActions.getDefaultFileAction(mimetype, 'file', OC.PERMISSION_READ);
+				if (action && action.action) {
 					action($('#filename').val());
```
Is this supposed to be `action.action($('#filename').val());`?

azul replied:
yes... thanks! fixed and renamed the actionSpec from `action` to `spec` so it becomes less confusing.

gary-kim reviewed (commented):
Noticed a couple more things.
* `Ctrl+click` on a directory in a public share doesn't go into that directory and opens a new tab with the same page.
* `Ctrl+click` on a file in a public share to be opened by the `text` app doesn't work as expected. This is more of an edge case though so I'd say it'd be okay to fix this in a follow up.

azul commented:
Found another issue...
If the file that the link points to has been moved from the folder it will trigger a download of the zipped folder content instead.

@jancborchardt what would the desired behavior be when:
* I create `MyFile.md`.
* I copy the link to `MyFile.md` in the folder (that is used to open it in another tab)
* I move `MyFile.md` to another place
* I click on the link.

I see a number of options:
1. open the same file in the new folder (we can tell where it is by the file id)
2. display a message that the file has been moved (maybe including the new place)
3. only display the folder

I guess 2. would be the right thing to do. But maybe i am missing something.

azul commented:
> Noticed a couple more things.
> 
>     * `Ctrl+click` on a directory in a public share doesn't go into that directory and opens a new tab with the same page.
> 
>     * `Ctrl+click` on a file in a public share to be opened by the `text` app doesn't work as expected. This is more of an edge case though so I'd say it'd be okay to fix this in a follow up.

I was able to confirm this and track it down a bit.
The call to `linkTo` is called from https://github.com/nextcloud/server/pull/20185/files#diff-dafefd9519924bec4efa9f5af4af8e68L1532 with the dir param as i would expect it.

However on public shares this is handled by 
```
# /apps/files_sharing/js/public.js
this.fileList.linkTo = function (dir) {
  return OC.generateUrl('/s/' + token + '', {dir: dir});
}
```

Turns out this completely ignores the dir parameter. I have no idea why.

@gary-kim do you have a clue?

azul commented:
58a87d0babcb91aab75b45e630d2fc2fee15691e was the last commit that touched the `linkTo` function in public.js.

azul commented:
> Noticed a couple more things.
> 
>     * `Ctrl+click` on a directory in a public share doesn't go into that directory and opens a new tab with the same page.
> 
>     * `Ctrl+click` on a file in a public share to be opened by the `text` app doesn't work as expected. This is more of an edge case though so I'd say it'd be okay to fix this in a follow up.

Okay... i figured it out and updated the pull request accordingly. The `linkTo` function in `public.js` had to be [fixed](https://github.com/nextcloud/server/pull/20185/files#diff-f2e13306e758275374ea68815da8af04L254).

azul commented:
I renamed the `editfile` parameter to `openfile` because depending on the permissions the file will also be shown read only if needed.

@gary-kim I would say this is ready to be merged. I'd be okay with adding more tests but i'd prefer to just see this included :smile: .

In terms of tests i think it would be nice to test that the generated urls actually do what they are supposed to. I really like that the generated url with `openfile` and `path` params on public shares work.

However i have no idea about how to write such a test. If you can point me to an example i'd also be happy to take it from there.

gary-kim reviewed (approved):
LGTM at this point
